package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,491 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#define _GNU_SOURCE
+
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <search.h>
+
+#include <libsbp/linux.h>
+#include <dirent.h>
+#include <libpiksi/logging.h>
+#include <libpiksi/util.h>
+#include <libpiksi/runit.h>
+#include <libpiksi/table.h>
+
+#include "query_sys_state.h"
+#include "resource_monitor.h"
+#include "resource_query.h"
+#include "resmon_common.h"
+
+//#define DEBUG_QUERY_FD_TAB
+
+#define ITEM_COUNT 10
+#define MAX_PROCESS_COUNT 1024
+#define MAX_FILE_COUNT 4096
+#define SBP_FRAMING_MAX_PAYLOAD_SIZE 255
+#define LSOF_CMD_BUF_LENGTH 1024 * 128
+typedef enum {
+  SEND_FD_COUNTS_0,
+  SEND_FD_COUNTS_1,
+  SEND_FD_COUNTS_2,
+  SEND_FD_COUNTS_3,
+  SEND_FD_COUNTS_4,
+  SEND_FD_COUNTS_5,
+  SEND_FD_COUNTS_6,
+  SEND_FD_COUNTS_7,
+  SEND_FD_COUNTS_8,
+  SEND_FD_COUNTS_9,
+  SEND_FD_SUMMARY,
+  SEND_DONE,
+} send_state_t;
+
+typedef struct {
+  char id_str[16];
+  u16 pid;
+  u16 fd_count;
+} pid_entry_t;
+
+typedef struct {
+  char file_str[128];
+  u16 fd_count;
+} file_entry_t;
+
+MAKE_TABLE_WRAPPER(pid, pid_entry_t)
+MAKE_TABLE_WRAPPER(file, file_entry_t)
+
+typedef struct {
+  send_state_t send_state;
+  u32 total_fd_count;
+  table_t *pid_table;
+  table_t *file_table;
+  const char *top10_pid_counts[ITEM_COUNT];
+  const char *top10_file_counts[ITEM_COUNT];
+} resq_state_t;
+
+typedef struct id_to_count {
+  u16 count;
+  const char *id;
+} id_to_count_t;
+
+enum {
+  FD_STATE_ID_STR = 0,
+  FD_STATE_EXE_FILE_STR,
+  FD_STATE_FD_STR,
+  FD_STATE_DONE,
+  FD_STATE_COUNT,
+};
+
+static bool is_real_file_path(const char *path)
+{
+  if (strncmp(path, "pipe", 4) == 0 || strncmp(path, "socket", 6) == 0)
+    return false;
+  else
+    return true;
+}
+
+static bool process_fd_entry(resq_state_t *state, const char *file_str, const char *id_str)
+{
+  pid_entry_t *entry = pid_table_get(state->pid_table, id_str);
+  state->total_fd_count++;
+  if (entry == NULL) {
+    pid_entry_t new_entry = {0};
+    int s = snprintf(new_entry.id_str, sizeof(entry->id_str), "%s", id_str);
+
+    if (s < 0 || (size_t)s >= sizeof(entry->id_str)) {
+#ifdef DEBUG_QUERY_FD_TAB
```

Comment:
Let's not disable this and make it a warning, this should never (or very infrequently) happen so it's safe to leave it enabled, let's also log the string that caused the problem and change